### PR TITLE
Generate configuration for launch.json

### LINKED ISSUE #3441: VsCode: generate configuration for debugging
This is a feature request. VS Code has built-in support for debugging through [CodeLLDB](https://marketplace.visualstudio.com/items?itemName=vadimcn.vscode-lldb) and [VS Code cpptools](https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools). These extensions can generate configuration for [launch.json](https://code.visualstudio.com/Docs/editor/debugging) through a menu that looks like this:

Slice 1

Getting the configuration right for Rust projects is a bit of a tricky one, with some nuance on how to properly target binaries, libraries, and tests across all major platforms. Since `rust-analyzer` is providing a premier Rust experience in VS Code, this might be an interesting feature to add.

## References

- I believe [this section](https://code.visualstudio.com/api/extension-guides/debugger-extension#using-a-debugconfigurationprovider) outlines how to add this functionality to VS Code extensions.
- [This post](https://jason-williams.co.uk/debugging-rust-in-vscode) shares a basic debugging setup for VS Code.

### LINKED ISSUE #4450: Environment variables when running tests
Rust Analyzer makes it really easy to run tests as you're authoring them in the editor - just using the "run test" action above the test name.

However, it would be really useful to be able to specify some environment variables when doing this - in particular, the RUST_LOG environment variable would allow for me to see logging output direct in VSCode when running these tests, whereas right now I need to go and run them in a terminal manually.

Cheers

### PR DESCRIPTION
**Pull request by vsrs**
This PR adds two new commands: `"rust-analyzer.debug"` and `"rust-analyzer.newDebugConfig"`. The former is a supplement to the existing `"rust-analyzer.run"` command and works the same way: asks for a runnable and starts new debug session. The latter allows adding a new configuration to **launch.json** (or to update an existing one).

If the new option `"rust-analyzer.debug.useLaunchJson"` is set to true then `"rust-analyzer.debug"` and Debug Lens will first look for existing debug configuration in **launch.json**. That is, it has become possible to specify startup arguments, env variables, etc.

`"rust-analyzer.debug.useLaunchJson"` is false by default, but it might be worth making true the default value. Personally I prefer true, but I'm not sure if it is good for all value.

----
I think that this PR also solves https://github.com/rust-analyzer/rust-analyzer/issues/3441.
Both methods to update launch.json mentioned in the issue do not work:
1. Menu. It is only possible to add a launch.json configuration template via a debug adapter. And anyway it's only a template and it is impossible to specify arguments from an extension.

2. DebugConfigurationProvider. The exact opposite situation: it is possible to specify all debug session settings, but it is impossible to export these settings to launch.json.

Separate `"rust-analyzer.newDebugConfig"` command looks better for me.

----
Fixes #4450
Fixes #3441

### CODE REVIEW

File: editors/code/package.json [hunk 1]
```diff
@@ -120,6 +120,16 @@
...
+            {
+                "command": "rust-analyzer.debug",
+                "title": "Debug",
+                "category": "Rust Analyzer"
+            },
+            {
+                "command": "rust-analyzer.newDebugConfig",
+                "title": "Generate launch configuration",
+                "category": "Rust Analyzer"
+            },
```
Comment (matklad) [inline editors/code/package.json:132]: Can we add a "save configuration" button to the quick pick? Looks like it supports some kind of extra buttons, tough it's not immediatelly clear how to use this funcitonality: https://code.visualstudio.com/api/references/vscode-api#QuickInputButton

File: editors/code/package.json [hunk 2]
```diff
@@ -433,6 +443,11 @@
                     "type": "object",
                     "default": {},
                     "description": "Optional settings passed to the debug engine. Example:\n{ \"lldb\": { \"terminal\":\"external\"} }"
+                },
+                "rust-analyzer.debug.useLaunchJson": {
```
Comment (matklad) [inline editors/code/package.json:447]: I'd say this should just always be `true`

File: editors/code/package.json (hunk 1 above)
Comment (vsrs) [reply editors/code/package.json:132]: Thanks, I didn't know about this API. Yes, we can. Though the button is not very noticeable:
save_cfg_button

I will add the button for both run and debug quick picks, but would prefer to leave the command too. At least to show the possibility in the commands list.

File: editors/code/package.json (hunk 2 above)
Comment (vsrs) [reply editors/code/package.json:447]: Do you mean that we do not need this option at all?

Comment (matklad) [reply editors/code/package.json:447]: Yes, I don’t think it is needed

File: editors/code/package.json (hunk 1 above)
Comment (matklad) [reply editors/code/package.json:132]: Seems reasonable!

While we are at it, maybe we should add `(from tasks.json)` to the label if we are loading persisted command

File: crates/rust-analyzer/src/main_loop/handlers.rs [hunk 3]
```diff
@@ -1011,14 +1011,17 @@ fn to_lsp_runnable(
...
     let (args, extra_args) = CargoTargetSpec::runnable_args(spec, &runnable.kind)?;
     let line_index = world.analysis().file_line_index(file_id)?;
     let label = match &runnable.kind {
         RunnableKind::Test { test_id, .. } => format!("test {}", test_id),
         RunnableKind::TestMod { path } => format!("test-mod {}", path),
         RunnableKind::Bench { test_id } => format!("bench {}", test_id),
         RunnableKind::DocTest { test_id, .. } => format!("doctest {}", test_id),
-        RunnableKind::Bin => "run binary".to_string(),
+        RunnableKind::Bin => {
+            target.map_or_else(|| "run binary".to_string(), |t| format!("run binary '{}'", t))
```
Comment (matklad) [inline crates/rust-analyzer/src/main_loop/handlers.rs:1023]:
```suggestion
            target.map_or_else(|| "run binary".to_string(), |t| format!("run {}", t))
```

?

File: editors/code/package.json (hunk 1 above)
Comment (vsrs) [reply editors/code/package.json:132]: The quick pick does not use any information from tasks.json, only from LSP. And I think we should leave it as it is. Or I did not get your idea.

**Comment by vsrs** [comment]
@matklad, what is the proper way to link issues to this PR? Merely add
> fixes #4450 
> fixes #3441

to the description? Looks like I cannot [link them manually](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#manually-linking-a-pull-request-to-an-issue). Or I shouldn't do it?

Sorry if this is a silly question, this is the first github project I'm contributing to.

**Comment by lnicola** [comment]
> what is the proper way to link issues to this PR?

Yup, GitHub recognizes a couple of keywords (I use "Fixes", "Closes" and "Implements", I don't think capitalization matters).

**Comment by vsrs** [comment]
@lnicola thanks. I'll try then to change the description and approve the PR.

**Pull request merged**